File: code_samples/section5/problems/problem5_4/problem5_4.py
"""Problem 5.4: Path Sum"""

from typing import Optional

# pylint: disable=too-few-public-methods


class TreeNode:
    """
    Model for a binary tree node.

    Each node stores:
    - val   : the integer value at this node
    - left  : reference to the left child (or None)
    - right : reference to the right child (or None)
    """
    def __init__(
        self,
        val: int = 0,
        left: "Optional[TreeNode]" = None,
        right: "Optional[TreeNode]" = None
    ):
        self.val = val
        self.left = left
        self.right = right


def has_path_sum(root: Optional[TreeNode], target_sum: int) -> bool:
    """
    Return True if there exists a ROOT-to-LEAF path whose values sum to target_sum.

    Important detail:
    - The path must end at a LEAF node (a node with no children).
    - It's not enough to match the sum at some internal node.

    Strategy (DFS recursion):
    - At each node, subtract node.val from target_sum to form a "remaining" sum.
    - Recurse down to children looking for a leaf that matches exactly.
    """
    # If we reached an empty subtree, it can't produce a valid path
    if root is None:
        return False

    # If this is a leaf node, check whether we've hit the target exactly
    if root.left is None and root.right is None:
        return root.val == target_sum

    # Otherwise, subtract this node's value and search children
    remaining = target_sum - root.val

    # If either subtree has a valid path with the remaining sum, return True
    return (
        has_path_sum(root.left, remaining)
        or has_path_sum(root.right, remaining)
    )


# ----------------------------
# Tree printer (sideways)
# ----------------------------

def print_tree(root: Optional[TreeNode]):
    """
    Print the tree sideways for visualization.

    Convention:
    - Right subtree prints first (appears "up")
    - Left subtree prints last (appears "down")
    """
    print("Tree (sideways, right is up):")
    _print_tree_impl(root, 0)
    print()


INDENT = 4  # number of spaces per depth level when printing


def _print_tree_impl(node: Optional[TreeNode], indent: int):
    """Recursive helper for sideways tree printing."""
    if node is None:
        return

    # Print right subtree first so it appears above the current node
    _print_tree_impl(node.right, indent + INDENT)

    # Print the current node with indentation based on depth
    print(" " * indent + str(node.val))

    # Print left subtree last so it appears below the current node
    _print_tree_impl(node.left, indent + INDENT)


# ----------------------------
# Test utility
# ----------------------------

def test_case(label: str, root: Optional[TreeNode], target: int):
    """
    Print a labeled test case:
    - shows the tree
    - prints the target sum
    - prints whether has_path_sum returns True/False
    """
    print(f"==== {label} ====")
    print_tree(root)
    print("Target sum:", target)
    print("has_path_sum:", has_path_sum(root, target))
    print()


# ----------------------------
# Build test trees
# ----------------------------

# 1. Empty tree (no paths exist)
empty = None

# 2. Single node tree (root is also a leaf)
single = TreeNode(7)

# 3. Large example tree:
#
#         5
#       /   \
#      4     8
#     /     / \
#    11    13  4
#   /  \        \
#  7    2        1
#
# This is a classic Path Sum example (LeetCode-style).
t = TreeNode(
    5,
    left=TreeNode(
        4,
        left=TreeNode(
            11,
            left=TreeNode(7),
            right=TreeNode(2)
        )
    ),
    right=TreeNode(
        8,
        left=TreeNode(13),
        right=TreeNode(
            4,
            right=TreeNode(1)
        )
    )
)

# Valid root-to-leaf path sums in this tree:
#   22 → 5 + 4 + 11 + 2
#   26 → 5 + 8 + 13
#   18 → 5 + 8 + 4 + 1
#
# Example of an invalid "sum" if you forget the leaf requirement:
#    5 → matches root alone, but root is NOT a leaf, so it must be False.

# ----------------------------
# Run tests
# ----------------------------

test_case("Empty tree (→ False)", empty, 10)
test_case("Single node (7 == 7 → True)", single, 7)
test_case("Single node (7 != 10 → False)", single, 10)

test_case("Large tree (path sum 22 → True)", t, 22)
test_case("Large tree (path sum 26 → True)", t, 26)
test_case("Large tree (path sum 18 → True)", t, 18)
test_case("Large tree (path sum 5 → False)", t, 5)

File: code_samples/section5/problems/problem5_4/problem5_4.cs
#nullable enable  // Enable nullable reference types (TreeNode? can be null)

// ==========================
// HAS PATH SUM (ROOT-TO-LEAF)
// ==========================

// Returns true if there exists a ROOT-to-LEAF path where the sum of node values
// along that path equals targetSum.
//
// Important: the path must end at a leaf node (a node with no children).
static bool HasPathSum(TreeNode? root, int targetSum)
{
    // Base case: empty subtree has no path
    if (root == null) return false;

    // If this is a leaf node, check if its value matches the remaining target sum
    if (root.Left == null && root.Right == null)
    {
        return root.Val == targetSum;
    }

    // Otherwise, subtract this node's value and continue searching
    int remaining = targetSum - root.Val;

    // If either left or right subtree contains a valid path, return true
    return HasPathSum(root.Left, remaining) ||
           HasPathSum(root.Right, remaining);
}

// ----------------------------
// Tree printer (sideways)
// ----------------------------

// Prints the tree sideways for visualization:
// - Right subtree appears above ("up")
// - Left subtree appears below ("down")
void PrintTree(TreeNode? root)
{
    Console.WriteLine("Tree (sideways, right is up):");
    PrintTreeImpl(root, 0);
    Console.WriteLine();
}

// Number of spaces to indent per tree level
const int INDENT = 4;

// Recursive helper for sideways printing
void PrintTreeImpl(TreeNode? node, int indent)
{
    // Base case: nothing to print
    if (node == null) return;

    // Print right subtree first so it appears "up"
    PrintTreeImpl(node.Right, indent + INDENT);

    // Print current node with indentation based on depth
    Console.WriteLine(new string(' ', indent) + node.Val);

    // Print left subtree last so it appears "down"
    PrintTreeImpl(node.Left, indent + INDENT);
}

// ----------------------------
// Test utility
// ----------------------------

// Runs a labeled test: prints the tree, target sum, and result
void TestCase(string label, TreeNode? root, int target)
{
    Console.WriteLine($"==== {label} ====");
    PrintTree(root);
    Console.WriteLine($"Target sum: {target}");
    Console.WriteLine("HasPathSum: " + HasPathSum(root, target));
    Console.WriteLine();
}

// ----------------------------
// Build test trees
// ----------------------------

// 1) Empty tree (always false for any target)
TreeNode? empty = null;

// 2) Single node tree
TreeNode single = new TreeNode(7);

// 3) Large test tree:
//
//         5
//       /   \
//      4     8
//     /     / \
//    11    13  4
//   /  \        \
//  7    2        1
TreeNode t = new(5)
{
    Left = new TreeNode(4)
    {
        Left = new TreeNode(11)
        {
            Left = new TreeNode(7),
            Right = new TreeNode(2)
        }
    },
    Right = new TreeNode(8)
    {
        Left = new TreeNode(13),
        Right = new TreeNode(4)
        {
            Right = new TreeNode(1)
        }
    }
};

// ----------------------------
// Run test cases
// ----------------------------

TestCase("Empty tree", empty, 10);
TestCase("Single node (7 == 7 → true)", single, 7);
TestCase("Single node (7 != 10 → false)", single, 10);

// Root-to-leaf path sums in 't':
// 5 + 4 + 11 + 7 = 27
// 5 + 4 + 11 + 2 = 22
// 5 + 8 + 13     = 26   (13 is a leaf in this tree)
// 5 + 8 + 4 + 1  = 18
TestCase("Large tree (has path sum 22 → true)", t, 22);
TestCase("Large tree (has path sum 26 → true)", t, 26);
TestCase("Large tree (has path sum 18 → true)", t, 18);

// This is false because 5 is not a leaf; path sums must end at a leaf node.
TestCase("Large tree (no path sum == 5 → false)", t, 5);

// ==========================
// TREE NODE DEFINITION
// ==========================

class TreeNode(int v)
{
    // Value stored in this node
    public int Val = v;

    // Child references (nullable because children may not exist)
    public TreeNode? Left;
    public TreeNode? Right;
}
